src/mono/mono/mini/mini-arm64.c

```diff
@@ -1529,15 +1529,26 @@ is_hfa (MonoType *t, int *out_nfields, int *out_esize, int *field_offsets)
 }
 
 static void
-add_valuetype (CallInfo *cinfo, ArgInfo *ainfo, MonoType *t)
+add_valuetype (CallInfo *cinfo, ArgInfo *ainfo, MonoType *t, gboolean is_return)
 {
 	int i, size, align_size, nregs, nfields, esize;
 	int field_offsets [16];
 	guint32 align;
+	MonoClass *klass;
 
+	klass = mono_class_from_mono_type_internal (t);
 	size = mini_type_stack_size_full (t, &align, cinfo->pinvoke);
 	align_size = ALIGN_TO (size, 8);
 
+	if (m_class_is_simd_type (klass) && !cinfo->pinvoke && !is_return && cinfo->fr < FP_PARAM_REGS) {
```

Comment:
I'm not sure if we are following the standard aarch64 calling convention, shouldn't we also return vector types in vector registers?

Reply:
Eventually, yes. That requires more work.